File: macosx/GlobalOptionsPopoverViewController.m
/******************************************************************************
 * Copyright (c) 2011-2012 Transmission authors and contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *****************************************************************************/

#import "GlobalOptionsPopoverViewController.h"

@implementation GlobalOptionsPopoverViewController

- (id) initWithHandle: (tr_session *) handle
{
    if ((self = [super initWithNibName: @"GlobalOptionsPopover" bundle: nil]))
    {
        fHandle = handle;

        fDefaults = [NSUserDefaults standardUserDefaults];
    }

    return self;
}

- (void) awakeFromNib
{
    [fUploadLimitField setIntValue: [fDefaults integerForKey: @"UploadLimit"]];
    [fDownloadLimitField setIntValue: [fDefaults integerForKey: @"DownloadLimit"]];

    [fRatioStopField setFloatValue: [fDefaults floatForKey: @"RatioLimit"]];
    [fIdleStopField setIntegerValue: [fDefaults integerForKey: @"IdleLimitMinutes"]];

    [[self view] setFrameSize: [[self view] fittingSize]];
}

- (IBAction) updatedDisplayString: (id) sender
{
    [[NSNotificationCenter defaultCenter] postNotificationName: @"RefreshTorrentTable" object: nil];
}

- (IBAction) setDownSpeedSetting: (id) sender
{
    tr_sessionLimitSpeed(fHandle, TR_DOWN, [fDefaults boolForKey: @"CheckDownload"]);

    [[NSNotificationCenter defaultCenter] postNotificationName: @"SpeedLimitUpdate" object: nil];
}

- (IBAction) setDownSpeedLimit: (id) sender
{
    const NSInteger limit = [sender integerValue];
    [fDefaults setInteger: limit forKey: @"DownloadLimit"];
    tr_sessionSetSpeedLimit_KBps(fHandle, TR_DOWN, limit);

    [[NSNotificationCenter defaultCenter] postNotificationName: @"UpdateSpeedLimitValuesOutsidePrefs" object: nil];
    [[NSNotificationCenter defaultCenter] postNotificationName: @"SpeedLimitUpdate" object: nil];
}

- (IBAction) setUpSpeedSetting: (id) sender
{
    tr_sessionLimitSpeed(fHandle, TR_UP, [fDefaults boolForKey: @"CheckUpload"]);

    [[NSNotificationCenter defaultCenter] postNotificationName: @"UpdateSpeedLimitValuesOutsidePrefs" object: nil];
    [[NSNotificationCenter defaultCenter] postNotificationName: @"SpeedLimitUpdate" object: nil];
}

- (IBAction) setUpSpeedLimit: (id) sender
{
    const NSInteger limit = [sender integerValue];
    [fDefaults setInteger: limit forKey: @"UploadLimit"];
    tr_sessionSetSpeedLimit_KBps(fHandle, TR_UP, limit);

    [[NSNotificationCenter defaultCenter] postNotificationName: @"SpeedLimitUpdate" object: nil];
}

- (IBAction) setRatioStopSetting: (id) sender
{
    tr_sessionSetRatioLimited(fHandle, [fDefaults boolForKey: @"RatioCheck"]);

    //reload main table for seeding progress
    [[NSNotificationCenter defaultCenter] postNotificationName: @"UpdateUI" object: nil];

    //reload global settings in inspector
    [[NSNotificationCenter defaultCenter] postNotificationName: @"UpdateGlobalOptions" object: nil];
}

- (IBAction) setRatioStopLimit: (id) sender
{
    const CGFloat value = [sender floatValue];
    [fDefaults setFloat: value forKey: @"RatioLimit"];
    tr_sessionSetRatioLimit(fHandle, value);

    [[NSNotificationCenter defaultCenter] postNotificationName: @"UpdateRatioStopValueOutsidePrefs" object: nil];

    //reload main table for seeding progress
    [[NSNotificationCenter defaultCenter] postNotificationName: @"UpdateUI" object: nil];

    //reload global settings in inspector
    [[NSNotificationCenter defaultCenter] postNotificationName: @"UpdateGlobalOptions" object: nil];
}

- (IBAction) setIdleStopSetting: (id) sender
{
    tr_sessionSetIdleLimited(fHandle, [fDefaults boolForKey: @"IdleLimitCheck"]);

    //reload main table for remaining seeding time
    [[NSNotificationCenter defaultCenter] postNotificationName: @"UpdateUI" object: nil];

    //reload global settings in inspector
    [[NSNotificationCenter defaultCenter] postNotificationName: @"UpdateGlobalOptions" object: nil];
}

- (IBAction) setIdleStopLimit: (id) sender
{
    const NSInteger value = [sender integerValue];
    [fDefaults setInteger: value forKey: @"IdleLimitMinutes"];
    tr_sessionSetIdleLimit(fHandle, value);

    [[NSNotificationCenter defaultCenter] postNotificationName: @"UpdateIdleStopValueOutsidePrefs" object: nil];

    //reload main table for remaining seeding time
    [[NSNotificationCenter defaultCenter] postNotificationName: @"UpdateUI" object: nil];

    //reload global settings in inspector
    [[NSNotificationCenter defaultCenter] postNotificationName: @"UpdateGlobalOptions" object: nil];
}

- (BOOL) control: (NSControl *) control textShouldBeginEditing: (NSText *) fieldEditor
{
    fInitialString = [control stringValue];

    return YES;
}

- (BOOL) control: (NSControl *) control didFailToFormatString: (NSString *) string errorDescription: (NSString *) error
{
    NSBeep();
    if (fInitialString)
    {
        [control setStringValue: fInitialString];
        fInitialString = nil;
    }
    return NO;
}

@end

File: macosx/GlobalOptionsPopoverViewController.swift
// This file Copyright © 2011-2023 Transmission authors and contributors.
// It may be used under the MIT (SPDX: MIT) license.
// License text can be found in the licenses/ folder.

import AppKit

class GlobalOptionsPopoverViewController: NSViewController {
    private let fSession: UnsafePointer<c_tr_session> = Transmission.cSession
    private let fDefaults = UserDefaults.standard

    @IBOutlet private weak var fUploadLimitField: NSTextField!
    @IBOutlet private weak var fDownloadLimitField: NSTextField!
    @IBOutlet private weak var fRatioStopField: NSTextField!
    @IBOutlet private weak var fIdleStopField: NSTextField!

    var fInitialString: String?

    init() {
        super.init(nibName: "GlobalOptionsPopover", bundle: nil)
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func awakeFromNib() {
        super.awakeFromNib()
        fUploadLimitField.integerValue = fDefaults.integer(forKey: "UploadLimit")
        fDownloadLimitField.integerValue = fDefaults.integer(forKey: "DownloadLimit")
        fRatioStopField.floatValue = fDefaults.float(forKey: "RatioLimit")
        fIdleStopField.integerValue = fDefaults.integer(forKey: "IdleLimitMinutes")

        view.setFrameSize(view.fittingSize)
    }

    @IBAction private func updatedDisplayString(_ sender: Any) {
        NotificationCenter.default.post(name: NSNotification.Name("RefreshTorrentTable"), object: nil)
    }

    @IBAction private func setDownSpeedSetting(_ sender: NSButton) {
        c_tr_sessionLimitSpeed(fSession, C_TR_DOWN, fDefaults.bool(forKey: "CheckDownload"))

        NotificationCenter.default.post(name: NSNotification.Name("SpeedLimitUpdate"), object: nil)
    }

    @IBAction private func setDownSpeedLimit(_ sender: NSTextField) {
        let limit = sender.integerValue
        fDefaults.set(limit, forKey: "DownloadLimit")
        c_tr_sessionSetSpeedLimit_KBps(fSession, C_TR_DOWN, limit)

        NotificationCenter.default.post(name: NSNotification.Name("UpdateSpeedLimitValuesOutsidePrefs"), object: nil)
        NotificationCenter.default.post(name: NSNotification.Name("SpeedLimitUpdate"), object: nil)
    }

    @IBAction private func setUpSpeedSetting(_ sender: NSButton) {
        c_tr_sessionLimitSpeed(fSession, C_TR_UP, fDefaults.bool(forKey: "CheckUpload"))

        NotificationCenter.default.post(name: NSNotification.Name("UpdateSpeedLimitValuesOutsidePrefs"), object: nil)
        NotificationCenter.default.post(name: NSNotification.Name("SpeedLimitUpdate"), object: nil)
    }

    @IBAction private func setUpSpeedLimit(_ sender: NSTextField) {
        let limit = sender.integerValue
        fDefaults.set(limit, forKey: "UploadLimit")
        c_tr_sessionSetSpeedLimit_KBps(fSession, C_TR_UP, limit)

        NotificationCenter.default.post(name: NSNotification.Name("SpeedLimitUpdate"), object: nil)
    }

    @IBAction private func setRatioStopSetting(_ sender: NSButton) {
        c_tr_sessionSetRatioLimited(fSession, fDefaults.bool(forKey: "RatioCheck"))

        // reload main table for seeding progress
        NotificationCenter.default.post(name: NSNotification.Name("UpdateUI"), object: nil)

        // reload global settings in inspector
        NotificationCenter.default.post(name: NSNotification.Name("UpdateGlobalOptions"), object: nil)
    }

    @IBAction private func setRatioStopLimit(_ sender: NSTextField) {
        let value = sender.doubleValue
        fDefaults.set(value, forKey: "RatioLimit")
        c_tr_sessionSetRatioLimit(fSession, value)

        NotificationCenter.default.post(name: NSNotification.Name("UpdateRatioStopValueOutsidePrefs"), object: nil)

        // reload main table for seeding progress
        NotificationCenter.default.post(name: NSNotification.Name("UpdateUI"), object: nil)

        // reload global settings in inspector
        NotificationCenter.default.post(name: NSNotification.Name("UpdateGlobalOptions"), object: nil)
    }

    @IBAction private func setIdleStopSetting(_ sender: NSButton) {
        c_tr_sessionSetIdleLimited(fSession, fDefaults.bool(forKey: "IdleLimitCheck"))

        // reload main table for remaining seeding time
        NotificationCenter.default.post(name: NSNotification.Name("UpdateUI"), object: nil)

        // reload global settings in inspector
        NotificationCenter.default.post(name: NSNotification.Name("UpdateGlobalOptions"), object: nil)
    }

    @IBAction private func setIdleStopLimit(_ sender: NSTextField) {
        let value = sender.intValue
        fDefaults.set(value, forKey: "IdleLimitMinutes")
        c_tr_sessionSetIdleLimit(fSession, UInt16(clamping: value))

        NotificationCenter.default.post(name: NSNotification.Name("UpdateIdleStopValueOutsidePrefs"), object: nil)

        // reload main table for remaining seeding time
        NotificationCenter.default.post(name: NSNotification.Name("UpdateUI"), object: nil)

        // reload global settings in inspector
        NotificationCenter.default.post(name: NSNotification.Name("UpdateGlobalOptions"), object: nil)
    }
}

extension GlobalOptionsPopoverViewController: NSControlTextEditingDelegate {
    func control(_ control: NSControl,
                 textShouldBeginEditing fieldEditor: NSText) -> Bool {
        fInitialString = control.stringValue
        return true
    }

    func control(_ control: NSControl,
                 didFailToFormatString string: String,
                 errorDescription error: String?) -> Bool {
        NSSound.beep()
        if let fInitialString = fInitialString {
            control.stringValue = fInitialString
            self.fInitialString = nil
        }
        return false
    }
}
